### DISCUSSION TITLE
Set Color from converter in theme friendly way

### DISCUSSION THREAD

**Discussion by nixtar** [AvaloniaUI/Avalonia#13664] [Q&A]
Hello,

I'm porting a WPF app to Avalonia and there is a converter that is setting the color of some text based on the value of an enum.

EG:
```c#
public object Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
{
    if (!(value is Status status)) return BindingOperations.DoNothing;

    switch (status)
    {
        case Status.Connecting:
        case Status.Disconnecting:
        case Status.Initialising:
            return new SolidColorBrush(Color.FromRgb(255, 141, 10));
        case Status.Connected:
            return new SolidColorBrush(Color.FromRgb(27, 153, 15));
        case Status.Disconnected:
            return new SolidColorBrush(Color.FromRgb(255, 51, 51));
        default:
            throw new ArgumentOutOfRangeException();
    }
}
```

This works fine however I'd like to do this in a theme friendly way.

One idea I had was to define the colors in a theme dictionary with the dark/light keys and have the converter to return the appropriate key.
However, I'm not sure how to get a converter to set a DynamicResource key?

Open to suggestions if there is a better way to handle this in general. 😄

**Comment by nixtar** [comment]
This is giving me some ideas:
https://github.com/carina-studio/ULogViewer/blob/master/ULogViewer/Converters/LogLevelBrushConverter.cs

**Comment by timunie** [comment]
you could use App.Current.TryFindResouce and return a given resource for example.

**Reply by nixtar** [reply]
What's a nice way to pass through the ThemeVariant for the context of the view/control?
I can access it from App.Current.ActualThemeVariant for the whole app. (works for me currently)
But if I ever wanted to do theme variants at a more granular level, I assume I'd need a way to pass it through as a param to the converter?

Theme change also appears to be an issue with this approach.
I could wire up my view models to fire a change notification on theme change but that feels a little bit hacky.

This was one of the reasons I was hoping to be able to wire up a DynamicResource with a converter for the key.

**Reply by timunie** [reply]
well check docs about resources pls. You can return an resource observable. 
https://docs.avaloniaui.net/docs/next/guides/styles-and-resources/resources#consuming-resources-from-code

**Reply by nixtar** [reply]
Sorry if I'm missing something obvious but I don't follow how that would work in my use case.
Is there a way I can return an observable from the converter?

My issue is that I want to change the Foreground of a single control based on the value of an enum in the view model.

**Reply by timunie** [reply]
maybe a custom Markup is needed, I'm not sure if it works from a converter other than a MulitConverter.

**Comment by nixtar** [comment]
This feels a bit hacky but I've come up with a possible solution for this.

To simplify things, I'll use an example only using a bool property `Online` instead of the enum.

I ended up using `MultiBinding` and a converter with one of the bindings being a DynamicResource thats part of the ThemeVariant.
This causes the MultiBinding and thus the converter to fire when the DynamicResource changes (ThemeVariant changes).

I'm also passing down the ActualThemeVariant of any ThemeVariantScopes.

EG:
```axaml
<MultiBinding Converter="{StaticResource OnlineStatusToColorConverter}">
  <Binding Path="Online"/>
  <Binding Path="ActualThemeVariant" RelativeSource="{RelativeSource Mode=FindAncestor, AncestorType=ThemeVariantScope}"/>
  <DynamicResource ResourceKey="TextControlForeground" />
</MultiBinding>
```

Then in the converter I'm able to try and account for all possible states:
```c#
public class OnlineStatusToColorConverter : IMultiValueConverter
{
    public object? Convert(IList<object?> values, Type targetType, object? parameter, CultureInfo culture)
    {
        if (values[0] is not bool online || targetType != typeof(IBrush)) return BindingOperations.DoNothing;
        
        var themeKey = online ? "Online" : "Offline";
        // Use supplied ThemeVariant
        if (values[1] is not ThemeVariant themeVariant)
        {
            // Use Application ThemeVariant if one is not supplied
            themeVariant = Application.Current!.ActualThemeVariant;
        }
        
        if (Application.Current!.TryFindResource(themeKey, themeVariant, out Object? output) && output is IBrush brush)
        {
            return brush;
        }
        
        return BindingOperations.DoNothing;
    }
}
```

End results:

https://github.com/AvaloniaUI/Avalonia/assets/2175630/808b5e57-686b-4797-86d4-2a00d59fc5a1

This doesn't account for all possible ThemeVariant sources however. EG the Window.
You could add these into the MultiBinding but it ends up getting messy fast.

One solution would be to bind to the control itself eg:

```axaml
      <MultiBinding Converter="{StaticResource OnlineStatusToColorConverter}">
        <Binding Path="Online"/>
        <Binding Path="ActualThemeVariant" RelativeSource="{RelativeSource Mode=Self}"/>
        <DynamicResource ResourceKey="TextControlForeground" />
      </MultiBinding>
```

However, this fails to build with error:
`Error AVLN:0004 Avalonia: Unable to resolve property or method of name 'ActualThemeVariant' on type 'Avalonia.Data.MultiBinding'. Line 64, position 24.`

Using `<Binding Path="$parent[TextBox].ActualThemeVariant"/>` or `<Binding Path="ActualThemeVariant" RelativeSource="{RelativeSource Mode=FindAncestor, AncestorType=TextBlock}"/>` builds but always returns `UnsetValueType`.

Ahh after typing this out I found #13528 using the workaround mentioned there and using `<Binding Path="ActualThemeVariant" ElementName="TestTextBlock"/>` works 😄.
